### PR TITLE
support deepspeed

### PR DESCRIPTION
**Pull request by BootsofLagrangian**
## Introduction

This PR adds [DeepSpeed ](https://github.com/microsoft/DeepSpeed)support via [Accelerate](https://huggingface.co/docs/accelerate/index) to sd-scripts, aiming to improve multi-GPU training with [ZeRO-Stage](https://www.deepspeed.ai/tutorials/zero/). I've made these changes in my fork under the `branch-deepspeed` and I'm open to any feedback!

### 1. Install DeepSpeed
First, activate your virtual environment and install DeepSpeed with the following command:

`DS_BUILD_OPS=0 pip install deepspeed`

### 2. Configure Accelerate
You can easily set up your environment for DeepSpeed with accelerate config. It allows you to control basic DeepSpeed environment variables. You can also use command-line arguments for configuration. Here's how you can set up for ZeRO-2 stage using Accelerate:

```
(deepspeed) accelerate config
In which compute environment are you running? **This machine**
Which type of machine are you using? **multi-GPU**
How many different machines will you use (use more than 1 for multi-node training)? [1]: **1**
Should distributed operations be checked while running for errors? This can avoid timeout issues but will be slower. [yes/NO]: **NO**
Do you wish to optimize your script with torch dynamo?[yes/NO]: **NO**
Do you want to use DeepSpeed? [yes/NO]: **yes**
Do you want to specify a json file to a DeepSpeed config? [yes/NO]: **NO**
What should be your DeepSpeed's ZeRO optimization stage? **2**
How many gradient accumulation steps you're passing in your script? [1]: **1**
Do you want to use gradient clipping? [yes/NO]: **NO**
How many GPU(s) should be used for distributed training? [1]: **4**
Do you wish to use FP16 or BF16 (mixed precision)? **bf16**
accelerate configuration saved at ~/.cache/huggingface/accelerate/default_config.yaml
```

Follow the prompts to select your environment settings, including using multi-GPU, enabling DeepSpeed, and setting ZeRO optimization stage to 2.

Your configuration will be saved in a YAML file, similar to the following example (path and values may vary):

```
compute_environment: LOCAL_MACHINE
debug: false
deepspeed_config:
  gradient_accumulation_steps: 1
  zero_stage: 2
...
```

### 3. Use in Your Scripts
TOML Configuration File: Add `deepspeed=true` and `zero_stage=[your_stage]` to your TOML config file. Refer to the ZeRO-stage and Accelerate DeepSpeed documentation for more details.
Bash Argument: Simply add `--deepspeed --zero_stage=[your_stage]` to your script's command line arguments.

### Note
This PR aims to improve training efficiency in multi-GPU setups. It's been tested only in Linux environments and specifically for multi-GPU configurations. The DeepSpeed support in Accelerate is still experimental, so please keep this in mind and feel free to provide feedback or comments on this PR.

### CODE REVIEW

**Comment by FurkanGozukara** [comment]
hello.

so we add `--deepspeed --zero_stage=[your_stage]`

but what is your_stage here? thank you

what is zero_stage?

**Comment by BootsofLagrangian** [comment]
> hello.
> 
> so we add `--deepspeed --zero_stage=[your_stage]`
> 
> but what is your_stage here? thank you
> 
> what is zero_stage?

your stage means one of ZeRO stage, [0, 1, 2, 3]. Details are in [ZeRO documents](https://www.deepspeed.ai/tutorials/zero/). 
Shortly
1. ZeRO stage 1 : The optimizer states
2. ZeRO stage 2 : The optimizer states + The gradient stages
3. ZeRO stage 3 : The optimizer states + The gradient stages + The parameters
to shard them into multi-gpus.

**Comment by FurkanGozukara** [comment]
> > hello.
> > so we add `--deepspeed --zero_stage=[your_stage]`
> > but what is your_stage here? thank you
> > what is zero_stage?
> 
> your stage means one of ZeRO stage, [0, 1, 2, 3]. Details are in [ZeRO documents](https://www.deepspeed.ai/tutorials/zero/). Shortly
> 
> 1. ZeRO stage 1 : The optimizer states
> 2. ZeRO stage 2 : The optimizer states + The gradient states
> 3. ZeRO stage 3 : The optimizer states + The gradient states + The parameters
> 
> to shard them into multi-gpus.

for 2 gpus which one you suggest? like --deepspeed --zero_stage=1?

what about 3 gpus?

currently it clone entire training on each gpus as far as i know

so can you tell some suggested guidelines?

**Comment by BootsofLagrangian** [comment]
> > > hello.
> > > so we add `--deepspeed --zero_stage=[your_stage]`
> > > but what is your_stage here? thank you
> > > what is zero_stage?
> > 
> > 
> > your stage means one of ZeRO stage, [0, 1, 2, 3]. Details are in [ZeRO documents](https://www.deepspeed.ai/tutorials/zero/). Shortly
> > 
> > 1. ZeRO stage 1 : The optimizer states
> > 2. ZeRO stage 2 : The optimizer states + The gradient states
> > 3. ZeRO stage 3 : The optimizer states + The gradient states + The parameters
> > 
> > to shard them into multi-gpus.
> 
> for 2 gpus which one you suggest? like --deepspeed --zero_stage=1?
> 
> what about 3 gpus?
> 
> currently it clone entire training on each gpus as far as i know
> 
> so can you tell some suggested guidelines?

I think that ZeRO-2 stage is optimal stage if you have enough VRAM. i.e. --deepspeed --zero_stage=2 only.

And if you want to use cpu/nvme offload, you might meet some kind of error in this commit. I will fix it soon.

But you can use ZeRO-2 stage without offloading

**Comment by FurkanGozukara** [comment]
when i tested fp16 and bf16 on SD 1.5 I had horrible results. are you able to get any decent results?

**Comment by mchdks** [comment]
any update here? @kohya-ss

**Comment by BootsofLagrangian** [comment]
Here is a report of deepspeed in sd-scripts

## Environment
### Model 
- Stable Diffusion XL 1.0

### GPUs
- GPUs, 24GB VRAM. **No NVLink**
  1. 2 x RTX 3090, 4 x RTX 3090, 6 x RTX 3090 and 8 x RTX 3090
  2. 2 x RTX 4090, 4 x RTX 4090 and 6 x RTX 4090

No NVLink make bottleneck among GPU communication.

### Requirements
1. torch 2.2.0 cu121
2. bitsandbytes 0.0.42 for AdamW8bit
3. accelerate 0.25.0
4. deepspeed 0.13.1
5. lycoris 2.0.2 for LoCon

### Training Settings
- Methodology
  1. Full Finetuning(**FT**). sdxl_train.py
  2. LoCon with rank=16(**PEFT**). sdxl_train_network.py
- Training Precision
  1. full_fp16, only active for **DDP**
- Distributed Method
  1. Distributed Data Parallel(**DDP**). sd-scripts and accelerate default
  2. **ZeRO**. stage 1, 2, and 3
- Resolution
  1. **1024x1024** with aspect ratio bucketing(ARB)
- Batch and Optimizer
  1. batch size = 4
  2. gradient accumulation steps 16
  3. AdamW8bit

With 24GB VRAM, sd-script can run **barely** PEFT, and can **not** run FT on **DDP**.

## Results
Fisrt, I'm sorry for some missing element of table. My budget is limited.

### RTX 3090, full_fp16, FT
VRAM usage(MB). Lower is Better.
| # of GPU | DDP | ZeRO-1 | ZeRO-2 | ZeRO-3 |
|:------:|:------:|:------:|:------:|:------:|
| 2 | OOM | OOM | OOM | OOM |
| 4 | OOM | OOM | OOM | 21539 |
| 6 | OOM | OOM | OOM | 21569 |
| 8 | OOM | OOM |  23400 | 21013 |

Training Speed(s/it). Lower is Better.
| # of GPU | DDP | ZeRO-1 | ZeRO-2 | ZeRO-3 |
|:------:|:------:|:------:|:------:|:------:|
| 2 | OOM | OOM | OOM | OOM |
| 4 | OOM | OOM | OOM | 191.73 |
| 6 | OOM | OOM | OOM | 477.72(?) |
| 8 | OOM | OOM | 143.58 | 190.09 |


### RTX 4090, full_fp16, FT
VRAM usage(MB). Lower is Better.
| # of GPU | DDP | ZeRO-1 | ZeRO-2 | ZeRO-3 |
|:------:|:------:|:------:|:------:|:------:|
| 2 | OOM | OOM | OOM | OOM |
| 4 | OOM | OOM | OOM | 21154 |
| 6 | OOM | OOM | OOM | 21571 |

Training Speed(s/it). Lower is Better.
| # of GPU | DDP | ZeRO-1 | ZeRO-2 | ZeRO-3 |
|:------:|:------:|:------:|:------:|:------:|
| 2 | OOM | OOM | OOM | OOM |
| 4 | OOM | OOM | OOM | 50.02 |
| 6 | OOM | OOM | OOM | 104.65 |


### RTX 3090, full_fp16, PEFT
VRAM usage(MB). Lower is Better.
| # of GPU | DDP | ZeRO-1 | ZeRO-2 | ZeRO-3 |
|:------:|:------:|:------:|:------:|:------:|
| 2 | 23282 | 22097 | 22026 | 21045 | 21821 |
| 4 | 23113 | 23206 | 23154 | 20968 | 23043 |
| 6 |22764 | 23280 | 23034 | 21949 | 23115 |
| 8 | 23109 | 22582 | SIGABRT* | 21955 | 22700 |

SIGABRT raised. IDK

Training Speed(s/it). Lower is Better.
| # of GPU | DDP | ZeRO-1 | ZeRO-2 | ZeRO-3 |
|:------:|:------:|:------:|:------:|:------:|
| 2 | 58.64 | 51.7 | 56.25 | 114.71 | 56.26 | 
| 4 | 59.27 | 55.92 | 58.37 | 167.77 | 58.33 |
| 6 | 58.32 | 52.61 | 57.07 | 375.46 | 57.73 |
| 8 |53.08 | 48.88 | SIGABRT | 165.85 | 52.2 |

### RTX 4090, full_fp16, PEFT
VRAM usage(MB). Lower is Better.
| # of GPU | DDP | ZeRO-1 | ZeRO-2 | ZeRO-3 |
|:------:|:------:|:------:|:------:|:------:|
| 2 | 22450 | 22163 | 22175 | 20791 | 21976 |
| 4 | 22720 | 23059 | 23528 | 22209 | 23172 |
| 6 | 21767 | OOM* | 23167 | 22142 | 23147 |

Ideally this OOM should not be happened.

Training Speed(s/it). Lower is Better.
| # of GPU | DDP | ZeRO-1 | ZeRO-2 | ZeRO-3 |
|:------:|:------:|:------:|:------:|:------:|
| 2 | 30.33 | 29.23 | 29.76 | 66.43 | 30.46 |
| 4 | 32.36 | 29.42 | 31.17 | 75.34 | 31.05 |
| 6 | 32.91 | - | 31.91 | 112.83 | 31.25 |

**Comment by FurkanGozukara** [comment]
the only way to utilize multiple consumer GPU is i think cloning the training if you don't have pro GPUs.

**Comment by tinbtb** [comment]
>  full_fp16, only active for DDP

Why not use bf16 for these cards?

>  Training Speed(s/it). Lower is Better.
> 29.23

Could you please compare it with training on a single card? As far as I remember I get roughly the same speeds with just one card.

**Comment by BootsofLagrangian** [comment]
> > full_fp16, only active for DDP
> 
> Why not use bf16 for these cards?
> 
> > Training Speed(s/it). Lower is Better.
> > 29.23
> 
> Could you please compare it with training on a single card? As far as I remember I get roughly the same speeds with just one card.

full_bf16 and bf16 is on running.

For a single card, training speed is almost same as DDP, DDP is slight slower.

**Comment by mchdks** [comment]
> > > full_fp16, only active for DDP
> > 
> > 
> > Why not use bf16 for these cards?
> > > Training Speed(s/it). Lower is Better.
> > > 29.23
> > 
> > 
> > Could you please compare it with training on a single card? As far as I remember I get roughly the same speeds with just one card.
> 
> full_bf16 and bf16 is on running.
> 
> For a single card, training speed is almost same as DDP, DDP is slight slower.

The results look very promising, I can help you by providing you with different multi-gpu machines if it will help with your tests.

**Comment by FurkanGozukara** [comment]
what is your effective batch size?

this is cloned on each GPU? 

like 2 gpu means 4 * 16 * 2 ?

**Comment by BootsofLagrangian** [comment]
> > > > full_fp16, only active for DDP
> > > 
> > > 
> > > Why not use bf16 for these cards?
> > > > Training Speed(s/it). Lower is Better.
> > > > 29.23
> > > 
> > > 
> > > Could you please compare it with training on a single card? As far as I remember I get roughly the same speeds with just one card.
> > 
> > 
> > full_bf16 and bf16 is on running.
> > For a single card, training speed is almost same as DDP, DDP is slight slower.
> 
> The results look very promising, I can help you by providing you with different multi-gpu machines if it will help with your tests.

Thank your suggestion. But I think now-days Diffusion Models are not too big and necessary to run with multi-machines or multi-nodes.

**Comment by BootsofLagrangian** [comment]
> what is your effective batch size?
> 
> this is cloned on each GPU?
> 
> like 2 gpu means 4 * 16 * 2 ?

Effective Batch calculation in sd-scripts is on below

[effective batch] = [number of machine] x [number of gpus] x [train_batch_size] x [gradient_accumulation_steps]

For example, in my 2 gpu settings. 

effective batch = 1 machine x 2 gpus x 4 train batch size x 16 gradient accumulation steps
 = 1 x 2 x 4 x 16 = 128

**Comment by BootsofLagrangian** [comment]
@mchdks 

Here is a simple yet all about installation.

### installation
I recommend to use deepspeed anaconda envrionment. First, you need to clone my deepspeed branch.

0. install anaconda
   - I recommend python=3.10
1. install anaconda deepspeed env via [yml file](https://github.com/microsoft/DeepSpeed/blob/master/environment.yml). 
   - `conda create -n deepspeed --file=/path/to/YAL_FILE.yml`
2. move dir to sd-scripts deepspeed branch
   - `cd /path/to/deepspeed/branch`
3. activate anaconda environment
   - `conda activate deepspeed`
4. install pytorch and sd-scripts requirements.txt
   - conda install pytorch==2.2.0 torchvision==0.17.0 pytorch-cuda=12.1 -c pytorch -c nvidia
   - pip install -r requirements.txt
   - tested torch version is 2.2.0 but it can run on other torch version >= 2.0.1. 
5. install bitsandbytes, xformers, and lycoris
   - pip install bitsandbytes xformers lycoris_lora
6. install deepspeed
   - DS_BUILD_OPS=0 pip install deepspeed=0.13.1
8. config accelerate config
9. prepare test image set and script configuration file, like

> CONFIG_FILE.toml
> pretrained_model_name_or_path = "./training/base_model/sd_xl_base_1.0.safetensors"
> xformers = true
> deepspeed = true
> zero_stage = 1
> mixed_precision = "bf16"
> save_precision = "bf16"
> full_bf16 = true
> output_name = "full_bf16_ff_zero_1"
> output_dir = "./training/ds_test/model"
> train_data_dir = "./training/test_img"
> shuffle_caption = true
> caption_extension = ".txt"
> random_crop = true
> resolution = "1024,1024"
> enable_bucket = true
> bucket_no_upscale = true
> save_every_n_epochs = 1
> train_batch_size = 4
> max_token_length = 225
> max_train_epochs = 1
> max_data_loader_n_workers = 4
> persistent_data_loader_workers = true
> seed = 42
> gradient_checkpointing = true
> gradient_accumulation_steps = 16
> logging_dir = "./training/ds_test/logs"
> caption_separator = ". "
> noise_offset = 0.0357
> learning_rate = 1e-4
> unet_lr = 1e-4
> learning_rate_te1 = 5e-5
> learning_rate_te2 = 5e-5
> train_text_encoder = true
> max_grad_norm = 1.0
> optimizer_type = "AdamW8bit"
> save_model_as = "safetensors"
> optimizer_args = [ "weight_decay=1e-1", ]
> lr_scheduler = "constant_with_warmup"
> lr_warmup_steps = 340
> no_half_vae = true

10. run script like this.
```
accelerate launch --mixed_precision=bf16 \
        --num_processes=8 --num_machines=1 --multi_gpu \
        --main_process_ip=localhost --main_process_port=29555 \
        --num_cpu_threads_per_process=4 \
        ./sdxl_train.py --config_file=$CONFIG_FILE
```
CONFIG_FILE is a path to above toml file.

11. you will meet very long warnings and logs. but just ignore them, it doesn't effect on training.

**Comment by BootsofLagrangian** [comment]
> @BootsofLagrangian Hi! I've merged the PR to the new branch, and I'm refactoring a bit the code.
> 
> I have a question for DeepSpeedWrapper. In my understanding, Accelerate library does not support multiple models for deepspeed. So, we need to wrap multiple models into a single model.
> 
> If this is correct, when we pass the wrapper to `accelerator.accumulate`, the argument for `accumulate` might be the wrapper, instead of the list of models. Because `accumulate` takes the prepared model.
> 
> Therefore, `training_models = [ds_model]` might be ok. Is this correct?

Yep, that is correct. accelerate do something magical, accumulate method accepts accelerate-compatible Modules.

I tested `training_models = [ds_model]` and it works.

**Comment by storuky** [comment]
I got very optimistic results!
I have only 3x4090 in my PC and I was able to run Adam (not Adam8bit) optimizer on BF16 (without full_bf16) with batch size 16 and 1st text encoder training.
I used zero stage 2 and cpu offloading.
**Speed is incredible: 10s/it** 😱 
It took 123 GB of RAM.
Just to note: to run Adam with batch size 16 you need ~ 75GB VRAM (A100 or H100) and it performs with 4s/it.

@BootsofLagrangian you are a wizard!

**Comment by FurkanGozukara** [comment]
@storuky wow nice results
